bindings/rust/s2n-tls/src/pool.rs

```diff
@@ -117,6 +117,19 @@ impl<T: Pool> Pool for Arc<T> {
     }
 }
 
+/// A pool of Connections. Not a pool of Configs.
+///
+/// Instead of allocating and freeing new `Connection`s, this struct allows them
+/// to be "wiped" and reused.
+///
+/// Customers should run their own benchmarks to determine exactly how much connection
+/// reuse can save, but in our own benchmarks savings are moderate.
+/// - connection reuse: 804 ns
+/// - connection allocation + free: 2.606 µs
+/// - **Total Savings: 1.802 µs per connection**
+///
+/// Generally a handshake will take ~ 1 ms, so connection reuse enables a 0.2%
+/// handshake performance improvement.
```

Comment:
I'm not a fan of concrete numbers to discuss performance, especially in documentation. These numbers are very specific to how you setup your benchmark. Normally I'd recommend ditching the concrete numbers for relative percentages (like you've got 0.2%), but I'm still not sure how helpful that is. Connection reuse as a feature exists specifically for situations where allocations are expensive, so you'd have to run the benchmark in an environment like that for useful numbers.

Maybe just explain the use case (expensive allocation), discuss how it's not always beneficial, and potentially mention alternative ways of reducing the cost of allocations?

Reply:
My general goal is to communicate to customers that the probably don't need this unless they have good reason to believe otherwise. I'm very open to difference ways to accomplish that.

> Connection reuse as a feature exists specifically for situations where allocations are expensive

How should customers determine whether their allocations are expensive? My impression was always that connection reuse was just recommended because "allocations are always expensive" and "less allocation is better". Customers that I have talked to have largely shared that perspective. I think that including the benchmarks attaches a helpful sense of scale to the discussion.

> but I'm still not sure how helpful that is.

In discussions with customers (and [this comment](https://github.com/aws/s2n-tls/pull/4708#discussion_r1717395881)) the actual performance numbers were a surprise. I think it's common for customers to think "I care about performance, therefore I should use this `ConfigPool` implementation that is surely more efficient", but they don't actually stop to write a benchmark.

Since they are unlikely to write a benchmark, I think that including a simplistic benchmark is better than nothing.

I edited the comment to
- call out that this is intended for environment where allocation is very expensive.
- the benchmark numbers are from a simplistic benchmark

Reply:
> My general goal is to communicate to customers that the probably don't need this unless they have good reason to believe otherwise. I'm very open to difference ways to accomplish that.

Then my suggestion is just to say that directly. The numbers are both less direct and not necessarily accurate for older environments.

If customers believe that allocations are expensive, either 1) they're right, and your testing environment is not the same as their environment or 2) they're wrong, but that's a bigger discussion than belongs on ConfigPool. Either way, I think trying to convince customers worried about allocations not to be worried about allocations is not a good approach to the ConfigPool documentation.

Reply:
> Since they are unlikely to write a benchmark

Since you did write a benchmark, maybe you can point customers towards that?